### ISSUE TITLE
Feedback

### ISSUE THREAD

**Issue by sophiewigmore** [joshuatcasey/libdependency#14]
Hi! I wanted to surface some feedback I had when I was working on https://github.com/paketo-buildpacks/mri/issues/393.

While implementing dependency management changes for the `ruby` dependency, for my initial implementation I was not going to use the top-level libdependency functions like `GenerateMetadataFunc` and what not, because I wanted to do the exercise of implementing it myself across Ruby repos, so that I could possibly factor out some code in a different way than is organized here.

## Feedback
Regardless, I had wanted to use some of the more basic utilities that are here, such as the [`GetNewVersions` function](https://github.com/joshuatcasey/libdependency/blob/4fd708ca98e64a78f0404b013036353cec896833/retrieve/get_new_versions_for_id.go#L13-L39). There's no arguing with the need for that function. However, the fact that it returns a `versionology.VersionFetcherArray` instead of a slice of string versions was a barrier to entry. Sure I could look at the implementation of `VersionFetcherArray` and figure out the conversion, but it made it substantially harder to use. On a similar note, the fact that the function takes in a `GetAllVersionsFunc` is another big barrier to entry, since I hadn't implemented this function exactly according to spec. If it had taken in the list of all versions or something it would be easier to use without having to leverage all the other aspects of this library.

I had a similar experience trying to use the `FilterUpstreamVersionsByConstraints` because it takes in custom types like `VersionFetcherArray` and a `versionology.Constraint` rather than a `cargo.Constraint`. Again, I could totally do conversions to the right types, but its a barrier to entry that I have to do all this conversion to use any of the majorly helpful features of the library.

The TLDR here is that when using the library, it sort of felt like it is entirely orchestrated around the main retrieval functions, and if those aren't used, then it becomes difficult to use the other functionality.
